### PR TITLE
feat: add terrain slope calculation and use storage

### PR DESCRIPTION
**Pull request by martinnj-adsk**
Example on using storage to store an analysis result and then using that result after running the analysis.

It will calculate the terrain slope, and draw a png based on a threshold. 

Depends on 

- [x] https://github.com/spacemakerai/forma-embedded-view-sdk/pull/52
- [x] https://github.com/spacemakerai/extension-service/pull/61
- [x] https://github.com/spacemakerai/forma-extensions/pull/128

image

### CODE REVIEW

**Comment by knut-sverdrup-adsk** [comment]
Since this repo is public, we should probably remove the links to dependency PRs in private repos from the description.

**Comment by knut-sverdrup-adsk** [comment]
Could we add the screenshot from PR description to the README as well? I think it is nice for developers browsing the examples to get an idea of what it does through figures.

File: analyses/terrain-slope/src/app.tsx [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+};
+
+const DEFAULT_SETTINGS: Settings = {
+  steepnessThreshold: 25,
+};
+
+export const RESOLUTION = 1;
+
+export const CANVAS_NAME = "terrain slope";
+// function calculateTerrainSteepness();
```
Comment (knut-sverdrup-adsk) [inline analyses/terrain-slope/src/app.tsx]:
```suggestion
```

File: analyses/terrain-slope/src/components/Calculate.tsx [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+import { useCallback } from "preact/hooks";
+import { Forma } from "forma-embedded-view-sdk/auto";
+import { CANVAS_NAME, RESOLUTION } from "../app";
+
+type Props = {
+  steepnessThreshold: number;
+};
+
+function getMinMax(array: Float32Array) {
+  // return [Math.min(array)]
```
Comment (knut-sverdrup-adsk) [inline analyses/terrain-slope/src/components/Calculate.tsx]:
```suggestion
```

File: analyses/terrain-slope/src/components/Calculate.tsx [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+      width,
+      height,
+      maxSlope,
+      minSlope,
+      degreesToRadians(steepnessThreshold)
+    );
+    const position = {
+      x: ((maxX + minX) * RESOLUTION) / 2,
+      y: ((maxY + minY) * RESOLUTION) / 2,
+      z: 29,
```
Comment (knut-sverdrup-adsk) [inline analyses/terrain-slope/src/components/Calculate.tsx]: might be worth a comment? 😄

File: analyses/terrain-slope/src/app.tsx [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+export default function App() {
+  const [projectSettings, setProjectSettings] = useState<Settings>();
+
+  useEffect(() => {
+    Forma.storage.getTextItem({ key: "settings" }).then((res) => {
```
Comment (havard-hoiby-adsk) [inline analyses/terrain-slope/src/app.tsx]: This moved to `Forma.extensions.storage` right?

**Comment by havard-hoiby-adsk** [comment]
@martinnj-adsk : This is great! :) 

I'm a bit conflicted on where I would propose we take this next. I would love for these extensions to be published and maintained. For that you would need to do a UX iteration to make this more consumable by users. In that respect I think that having two ways of saving and have all those buttons is a bit to much. But in the context of an example for developers to showcase storage functionality it is great :)

I would suggest a few changes: 
 - Extract all storage related methods to `src/services/storage.ts`. This is what we want the users to look at in this example. So just having it be in a single file makes it easier to browse
 - On the UI side there are two things missing, what is Steepness threshold? and where is that color bar that shows me what the colors mean? This first one would be solved by adding a unit (e.g. degrees), the second one is possible, but you need to look at how predictive is using the designmode API and then expose those APIs.

Both of these comments are non-blocking for merging. I suggest you just merge now, then we'll add the degree, and I could do the refactor pass on `storage`. The Color bar is a larger thing, that we need to discuss when to prioritize 👍

**Review by havard-hoiby-adsk — APPROVED** [review]

**Comment by martinnj-adsk** [comment]
> I would love for these extensions to be published and maintained. For that you would need to do a UX iteration to make this more consumable by users. 

Great point. I haven't spent any time on UX with these, so could definitely be better 😅 

> Extract all storage related methods to src/services/storage.ts. This is what we want the users to look at in this example. So just having it be in a single file makes it easier to browse

I think this makes sense right away, and a pretty quick fix, so I think I'll just refactor and then merge :) Then we can discuss the color bar later :)

**Pull request merged**